### ISSUE TITLE
`formatQuery` with `parseNumbers: true` converts comma-separated value string to an integer.

### ISSUE THREAD

**Issue by bborn** [react-querybuilder/react-querybuilder#745]
### React Query Builder version

v7.6.0

### Description

See screenshot. When `parseNumbers` is true, numeric between values get converted to an integer, losing the comma separator.

![CleanShot 2024-08-01 at 10 27 17@2x](https://github.com/user-attachments/assets/a3958c8e-bf72-4bc1-9603-5d0974b96841)

### Reproduction

Repro [using RQB demo page.](https://react-querybuilder.js.org/demo#s=eJylVF1r3DgU%2FSuLn0dgfdlSoIV8kFJoE3ZT2ofShyvp2qNdx5packMo%2Be8rj50ZZ0pblsXGxudK95x7rq6%2FF19HHB6Ls%2B%2BFd8VZUVMtKQdGaAmGCEEVMWBLIji1JeONQ1cWm8KGe%2BN7SGHIe6B3GepDKs4a6CJuimHsMBZnn5%2BTmqZEkCXRyFRO6nJSVzFiLCpdo8SKqZyh8dhNy6HF%2FBV2OCwEBtMDYp%2FBb9CNmBGmNpIVT1%2BeNrP%2Bt5f%2FtYJfyRU0X0ITV3NLhIWa5AIYKR3lFhhToqlXchs%2FxHQD97jSd5fw22kNre%2FjJ5%2B2RRa9t2yhayreKCkpYShqIrC2mU5bIstSi0ZQ2hi9ouvgB7aP4Dd%2FfISx3UL%2FktX3p2z%2FtxevX3ThNHu2ByUIIFLInL1qKIHGMKIrici1BFfa4gfDUVKsnGqItk5kTXmrVtYSJ6XUylQ1ZW6lycf3Y%2FTWnxb76igtDSNmaRk%2B1F0BaFUrUhqYOCgjKhtBHKsUNwxMxas1Rx9zjnvs0085ijejTzDMp3BtgrS1MooC4WgkEboxRAGTRAsmpHPUUl6uqNohjDt019Mn%2FTnbepl4xu%2FCOFjcH8Ip2YkQnWeBaZ0PslCaCEBOcosV4azkxjWc8xpXQowf0tZB%2Bu305QkTebxIyTdUV%2Fld5Xv2IeySD32cpjFuw8Pl4T8RL%2BY0f82tXyZvWnMT0ofQth2%2BQC%2B70OPFmNI%2B2yrwLth%2FTvEBI6bbfu%2FNZR6CdjkCh8jtUs5zcNkHYwp32KFN17MH86YjfHuwYYk4NxXwIdzgw5upIUcN2Rw%2Fmffn%2FD9dUN873GF%2B9GnlxSHc%2BZjieTwfBng8otiD6fBqgPa8d1dD2B0izscp5I4AmrF9H9yxph0MEW%2FGe4MTzyz77zEmn9vs3sFjGNMLPy8G6O32pCd3W9%2Bkc7v0ch%2FIzY3pcepSJm1g7FLx9C9NxPdt)

### Expected behavior

When `parseNumbers` is true, the comma separator should be respected.

Should be like this: https://share.cleanshot.com/YpDbYDFy

**Comment by jakeboone02** [comment]
Good catch. I'm wondering if `formatQuery` should automatically convert lists of numbers to arrays when `parseNumbers` is `true`.

For example, given the following query:

```ts
const query = {
  combinator: "and",
  rules: [{ field: "age", operator: "between", value: "28,52" }]
}
```

...what should be the result of this:

```ts
formatQuery(query, { format: 'json', parseNumbers: true })
```

Option 1, leave as `string` (as you suggested):

```diff
 {
   "combinator": "and",
   "rules": [
     {
       "field": "age",
       "operator": "between",
+      "value": "28,52"
     }
   ]
 }
```

Option 2, convert to `Array<number>`:

```diff
 {
   "combinator": "and",
   "rules": [
     {
       "field": "age",
       "operator": "between",
+      "value": [28, 52]
     }
   ]
 }
```

**Comment by bborn** [comment]
I'd say if `parseNumbers: true`, I'd expect  option 2 (`Array<number>`). Right now I'm having to split on the comma, so I think just getting an array would be nicer.

**Comment by jakeboone02** [comment]
Ok, I'll put up a PR that does that and release a patch once you've tested it.

**Comment by jakeboone02** [comment]
Let me know if [this](https://deploy-preview-748--react-querybuilder.netlify.app/demo#s=eJylVF1r3DgU%2FSuLn0dgfdlSoIV8kFJoE3ZT2ofShyvp2qNdx5packMo%2Be8rj50ZZ0pblsXGxudK95x7rq6%2FF19HHB6Ls%2B%2BFd8VZUVMtKQdGaAmGCEEVMWBLIji1JeONQ1cWm8KGe%2BN7SGHIe6B3GepDKs4a6CJuimHsMBZnn5%2BTmqZEkCXRyFRO6nJSVzFiLCpdo8SKqZyh8dhNy6HF%2FBV2OCwEBtMDYp%2FBb9CNmBGmNpIVT1%2BeNrP%2Bt5f%2FtYJfyRU0X0ITV3NLhIWa5AIYKR3lFhhToqlXchs%2FxHQD97jSd5fw22kNre%2FjJ5%2B2RRa9t2yhayreKCkpYShqIrC2mU5bIstSi0ZQ2hi9ouvgB7aP4Dd%2FfISx3UL%2FktX3p2z%2FtxevX3ThNHu2ByUIIFLInL1qKIHGMKIrici1BFfa4gfDUVKsnGqItk5kTXmrVtYSJ6XUylQ1ZW6lycf3Y%2FTWnxb76igtDSNmaRk%2B1F0BaFUrUhqYOCgjKhtBHKsUNwxMxas1Rx9zjnvs0085ijejTzDMp3BtgrS1MooC4WgkEboxRAGTRAsmpHPUUl6uqNohjDt019Mn%2FTnbepl4xu%2FCOFjcH8Ip2YkQnWeBaZ0PslCaCEBOcosV4azkxjWc8xpXQowf0tZB%2Bu305QkTebxIyTdUV%2Fld5Xv2IeySD32cpjFuw8Pl4T8RL%2BY0f82tXyZvWnMT0ofQth2%2BQC%2B70OPFmNI%2B2yrwLth%2FTvEBI6bbfu%2FNZR6CdjkCh8jtUs5zcNkHYwp32KFN17MH86YjfHuwYYk4NxXwIdzgw5upIUcN2Rw%2Fmffn%2FD9dUN873GF%2B9GnlxSHc%2BZjieTwfBng8otiD6fBqgPa8d1dD2B0izscp5I4AmrF9H9yxph0MEW%2FGe4MTzyz77zEmn9vs3sFjGNMLPy8G6O32pCd3W9%2Bkc7v0ch%2FIzY3pcepSJm1g7FLx9C9NxPdt) works as expected. You can also [create a sandbox](https://codesandbox.io/s/react-query-builder-basic-typescript-template-6rk8qj) with the PR code.

**Comment by bborn** [comment]
Yup - that looks great to me.

**Comment by jakeboone02** [comment]
Fixed in [v7.6.1](https://github.com/react-querybuilder/react-querybuilder/releases/tag/v7.6.1)

**Comment by bborn** [comment]
Awesome. Thanks!

**Comment by bborn** [comment]
@jakeboone02 oops, spoke too soon. I discovered another issue:

https://share.cleanshot.com/dBDG5QSl

When using a `between` operator, the value should always be an array, even if only one value exists. 

So if I do `between 0 and null` the value should be `[0,null]`, and if I do `between null and 1` the value should be `[null, 1]`.

Otherwise you can see what happens in the UI: https://share.cleanshot.com/FQd5FyGJ

Because the value doesn't account for an empty input, it resets the first between input to the value of the second one.

Hope that makes sense - LMK if I can clarify.

Thanks!

**Comment by jakeboone02** [comment]
Yeah, I see what you're saying. I think that's actually two different issues, but I'll address them both in another PR.

**Comment by jakeboone02** [comment]
@bborn I resolved the issue with the "between" value array collapsing into a single value in #769 (tracked in #760).

[Please test the changes here](https://deploy-preview-769--react-querybuilder.netlify.app/demo) and let me know if it works for your case.

Also, I made a somewhat related change to the `parseNumbers` prop options and I'd like your feedback on it. From the changelog:

> - The `parseNumbers` prop now accepts an optional "-limited" suffix on existing `string` config values "enhanced", "strict", and "native". When the "-limited" suffix is used (e.g., `parseNumbers="enhanced-limited"`), values will only be parsed for numericity when the `inputType` of the value editor evaluates to `"number"`.

When you select the `parseNumbers` option on the demo ([like this](https://deploy-preview-769--react-querybuilder.netlify.app/demo#s=eJy1VGFr3DgQ%2FSuHP69AtmTJDvQgSWkptAl3Ke2Hch9G0sir1rG2ktwQSv77yWtn17ulBwcteO31jPTezHsjfy%2B%2Bjhgei4vvhTPFRVGzum2YKImgkhIuVEuULiWxrG4Yk5zWwhabQvt75QZIPuQ9MJgcGnwqLiz0ETdFGHuMxcWnBbSiYIRCSpi2hvDKtqQ1NRJZWsWU0SXXLCNYh%2F203LoQ0w3cY459g37ECaLJL36HYeFU2LkhfnRpWzxtFhpRlRKQUVKVFRDOG0Ma23DCdauglRVFWa9ooMNTzD%2BPfFXz9M%2FTZtbmzfX%2FVef3SnGX8Nt%2FiDHbsdBZKoxGKXOZLRJeM0sUA05ANKKhGoBrXNH18APbB3CbPz7A2G1hOGV1wznbLzRgMvwMnUvKBJYTptaEg%2BGkwaomIFtmsdRSmWlIzgRvuCltKxlBUynCpc6WQSlyTZZKm%2F9KthbcxXdjdNqdN%2FviWFoKI%2BbScvi5srLO2BWlRPC2Jly1nORfTVRjZH7yhgOsOYaYMe5xSD%2FlKF6PLkEo9lN4IjHVQvJJXdtmEWqhiDKmJrLizBojdIl0RdUFP%2B7QvJpey5%2BzrZfx5%2FidH4PG%2FRBOYD8UkhujXGWvbUN4qyVpdGWIoo3QmlVcwFpX5ULaGkh4PrnpAXFYlVK2NSclJZRtylbkp8jXrIPfJeeHOJ3GuPUP14dvULyaYf6erV9O3rTmxqf3vut6PIle937AqzGlPdoq8dbrL%2BfxgBHT7bDX5jofgm4ZgUPmdmnnObnsgzH5O%2BxRp1ezBvOmY%2Fj2IMOSMWZq4L2%2FwYfXkyHHGrI4bhLvr%2FlbvUTdYHCH%2BTaklRaHdO9iipfxMgR4PEZxANXjywDd5WBeBr87ZIyLU8ocA6jG7p03x552ECLejPcKJ5657M9jTC7bbN7Cox%2FTiZ5XAQa9PfPkbutsutSLl%2FtENjemx8mlTGph7FPx9C%2FJ4BJA)), the `"enhanced-limited"` option will be used. The query in that link has two rules, one with `inputType: "number"` (field "age") and the other with `inputType: "text"` (field "firstName"). You should be able to see on the JSON tab that the "Age" value is parsed to a `number` but the "First Name" value remains a `string`.

**Comment by bborn** [comment]
@jakeboone02 confirming the `between` fix works as I would expect in the demo: https://share.cleanshot.com/yCmhF8VL

Regarding the `-limited` change: I think it makes sense. I'm a bit thrown off by the naming, since the `parseNumbers` prop on `QueryBuilder` is really more of a validator (it ensures numericity of whatever you input into the value editor, according to a few different algorithms), while the `parseNumbers` option on `formatQuery` is more of a formatter (it converts string values to a number using a different algorithm). Not sure if there's anything to be done there, but just my two cents.

**Comment by jakeboone02** [comment]
I had the same thought about them being different things with the same name. To be honest I got them mixed up more than a few times when I started working on that PR.

FWIW, the "numericity" algorithm choices are the same for both options (enhanced/native/strict). But you're right that they serve different purposes. Somewhat overlapping, but not completely.

Anyway, I'll think about it a little more but probably merge it pretty soon with no major changes. Thanks for your feedback!
